### PR TITLE
Initial Commit of Dylan's S3 Upload Script

### PR DESCRIPTION
**Pull request by dack-fe**

### CODE REVIEW

**Comment by fe-wickham** [comment]
This script finds all the artifacts S3 needs from the last generated build and uploads the artifacts to S3

**Review by fe-wickham — APPROVED** [review]

File: autogen/s3upload.m [hunk 1]
```diff
@@ -0,0 +1,51 @@
+% s3upload(mp, bucket, path)
```
Comment (tvannoy) [inline autogen/s3upload.m]: `clean` isn't included in this doc comment

File: autogen/s3upload.m [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        system(cmd);
+    end
+    
+    paths = getArtifactPaths(mp);
+    for k=1:length(paths)
+        path = paths{k}; 
+        cmd = "aws s3 cp " + path + " " + s3path + "/";
+        system(cmd);
+    end
+    %;
```
Comment (tvannoy) [inline autogen/s3upload.m]: Not sure what this is doing here

**Review by tvannoy — CHANGES_REQUESTED** [review]

File: autogen/s3upload.m [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+% Audio Logic
+% 985 Technology Blvd
+% Bozeman, MT 59718
+% openspeech@flatearthinc.com
+
+function paths = s3upload(mp, bucket, path, clean)
+
+    s3path = 's3://' + string(bucket) + '/' + string(path);
+    disp(s3path)
+    if exist('clean', 'var') && clean
```
Comment (tvannoy) [inline autogen/s3upload.m]: What's the advantage of this conditional instead of just using `if clean` (assuming `clean` is a bool)?  @fe-wickham

**Comment by tvannoy** [comment]
Requested changes are just minor cleanup. Functionality looks :+1:

File: autogen/s3upload.m (hunk 3 above)
Comment (fe-tdavis) [reply autogen/s3upload.m]: It seems like clean is an optional argument.  Perhaps noting that somewhere would make this a little more clear. Correct me if I'm wrong, but this is basically saying if clean was passed in and true, then run a command, but if it wasn't passed in, do nothing?

**Review by fe-tdavis — APPROVED** [review]

File: autogen/s3upload.m (hunk 3 above)
Comment (tvannoy) [reply autogen/s3upload.m]: That makes sense. I've never seen that pattern for optional arguments before, but I rarely write or read matlab code with optional arguments. I've seen optional arguments handled like this before:
```matlab
if nargin < 4
    clean = false;
end if;
```
So an alternative could be to use `nargin` and assign a default value to clean when it is not included in the function call. That would make it so we don't have to check if `clean` exists. I don't know if either solution is much preferable to the other.

Comment (fe-tdavis) [reply autogen/s3upload.m]: I've typically gone with the `nargin` style as well, but I'm not sure if there's some sort of recommended standard for optional arguments.  The `&&` is pretty slick, so I'm good with leaving it as is.  :)

Comment (tvannoy) [reply autogen/s3upload.m]: I don't have a preference either way. The most popular [MATLAB style guidelines](https://www.mathworks.com/matlabcentral/fileexchange/46056-matlab-style-guidelines-2-0) don't say anything about optional arguments.

Mathworks has a page on [argument validation](https://www.mathworks.com/help/matlab/matlab_prog/function-argument-validation-1.html#mw_43018d69-1d4e-4d7a-813d-b25d64a9db15). They do mention the `nargin` approach, though they spend more time talking about using argument blocks, which are new to me. The argument block can specify the data types of arguments as well as default values for optional positional arguments; they can also handle name-value pair arguments, which are common in official MATLAB functions.

File: autogen/s3upload.m (hunk 2 above)
Comment (fe-wickham) [reply autogen/s3upload.m]: Its just hanging out

File: autogen/s3upload.m (hunk 3 above)
Comment (fe-wickham) [reply autogen/s3upload.m]: I have updated it to use the arguments block syntax which does appear to be pretty clean

File: autogen/s3upload.m (hunk 1 above)
Comment (fe-wickham) [reply autogen/s3upload.m]: Fixed

**Review by tvannoy — APPROVED** [review]

File: autogen/s3upload.m (hunk 3 above)
Comment (tvannoy) [reply autogen/s3upload.m]: I do like the arguments block as well. Wish I would have known about it before.

**Pull request merged**